### PR TITLE
Adds `dim` argument to `torch.unique`

### LINKED ISSUE #9997: [feature request] unique to work on nd tensors via dim arg
Can we have `unique` to return unique rows and cols as well, similar to `numpy.unique`? Example usage to give an idea:
```
>>> a = torch.tensor([[1, 0, 0], [1, 0, 0], [2, 3, 3]])
>>> torch.unique(a)
tensor([0, 1, 2, 3])
>>> torch.unique(a, dim=0)
tensor([[1, 0, 0], [2, 3, 3]])
>>> torch.unique(a, dim=1)
tensor([[1, 0], [1, 0], [2, 3]])
```

### PR DESCRIPTION
**Pull request by ptrblck**
Initial version of `unique` supporting a `dim` argument.

As discussed in [this issue](https://github.com/pytorch/pytorch/issues/9997) I added the `dim` argument to `torch.unique` with the same behavior like [numpy](https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.unique.html).

Since the implementation is based on `std/thrust::unique`, the `tensor` always needs to be sorted. The `sorted` argument in `torch.unique` does not have any function, as in the CUDA version of the plain `torch.unique`.

To check the performance and equal behavior between `torch.unique` and `np.unique`, I've used [this gist](https://gist.github.com/ptrblck/ac0dc862f4e1766f0e1036c252cdb105).

Currently we achieve the following timings for an input of `x = torch.randint(2, (1000, 1000))`:
(The values are calculated by taking the average of the times for both dimension)

| Device | PyTorch (return_inverse=False) | Numpy (return_inverse=False) | PyTorch (return_inverse=True) | Numpy (return_inverse=True) |
| --- | --- | --- | --- | --- |
| CPU | ~0.007331s | ~0.022452s | ~0.011139s | ~0.044800s |
| GPU | ~0.006154s | - | ~0.105373s | - |

Many thanks to @colesbury for the awesome mentoring and the valuable advices on the general implementation and performance issues!

### CODE REVIEW

File: aten/src/ATen/native/Unique.cpp [hunk 1]
```diff
@@ -47,6 +47,81 @@ std::tuple<Tensor, Tensor> _unique_cpu_template(
...
+    input_sorted[i] = input_flat[indices[i]];
+  }
+ 
+  // pre-calculate mask for inverse_indices
+  Tensor mask = at::empty(input_sorted.size(0), self.type().toScalarType(kLong));
+  mask[0] = 1;
+  int mask_idx = 1;
+
+  std::vector<Tensor> input_unbind = at::unbind(input_sorted, 0);
+  auto last = std::unique(input_unbind.begin(), input_unbind.end(), [&](Tensor a, Tensor b) {
```
Comment (colesbury) [inline aten/src/ATen/native/Unique.cpp]: I think that manually writing the equivalent loop for computing unique values and computing inverse_indices as part of that loop is going to be faster.

(As you pointed out earlier, that's not going to work for CUDA)

There's an example implementation of unique here:
https://en.cppreference.com/w/cpp/algorithm/unique

Comment (ptrblck) [reply aten/src/ATen/native/Unique.cpp]: Thanks for the suggestion. The `mask` used for the `inverse_indices` is already being pre-computed in the `std::unique` lambda function. 
I'll have a look at the implementation and will try to re-use the logic to manually write the loop.

Would you like to compute the `inverse_indices` as part of this loop without the mask as it's currently done?

Comment (colesbury) [reply aten/src/ATen/native/Unique.cpp]: Yes, without the mask and the cumsum call.

There are two reasons for this:
1) I think it will be faster and may be less code
2) The current implementation assumes a certain implementation of std::unique that may not be true. (That unique traverses input_unbind once in order). While this is probably true for all implementations, I don't think it's guaranteed by the standard.

File: aten/src/ATen/native/cuda/Unique.cu [hunk 2]
```diff
@@ -69,6 +70,92 @@ template <typename scalar_t>
...
+            return false;
+          }
+        }
+        return true;
+      });
+    input_sorted_indices.resize_(last - input_sorted_indices_ptr);
+    Tensor output = input_sorted.index_select(0, input_sorted_indices);
+
+    // // reshape back
+    std::vector<int64_t> new_sizes(orig_sizes.begin(), orig_sizes.end());
```
[the remark this answers is not in the document]
Comment (ptrblck) [reply aten/src/ATen/native/cuda/Unique.cu]: That won't probably work, as `orig_sizes` is already a `std::vector`. 
However that's a good catch nonetheless, since I can just manipulate `orig_sizes` and use it to resize `output`.
I will just drop `new_sizes` completely in the CPU and CUDA version.

File: aten/src/ATen/native/cuda/Unique.cu [hunk 3]
```diff
@@ -7,6 +7,7 @@
 #include <tuple>
 #include <thrust/unique.h>
 #include <thrust/sort.h>
+#include <thrust/device_vector.h>
```
Comment (ptrblck) [inline aten/src/ATen/native/cuda/Unique.cu]: Not used anymore. Will remove it in the next commit.

File: aten/src/ATen/native/cuda/Unique.cu (hunk 2 above)
[the remark this answers is not in the document]
Comment (colesbury) [reply aten/src/ATen/native/cuda/Unique.cu]: In that case you can just use the copy-constructor. (It's fine to make a copy here).

This will work:

```
std::vector<int64_t> new_sizes = orig_sizes;
```

I generally prefer this syntax:

```
auto new_sizes = std::vector<int64_t>(orig_sizes);
```

File: aten/src/ATen/native/Unique.cpp (hunk 1 above)
Comment (colesbury) [reply aten/src/ATen/native/Unique.cpp]: The CUDA version is fine: the CUDA lambda doesn't modify anything. (It doesn't have this `mask_idx` counter)

**Comment by ptrblck** [comment]
As there is no switch between `return_inverse=True/False`, since `inverse_indices` will be pre-computed in the unique implementation, the new CPU time is now approx. `0.00921` seconds.

**Review by colesbury — APPROVED** [review]
lgtm, thanks @ptrblck

**Review by gchanan — APPROVED** [review]
